torch/utils/_pytree.py

```diff
@@ -964,11 +964,16 @@ def unflatten(self, leaves: Iterable[Any]) -> PyTree:
         return unflatten_fn(child_pytrees, self.context)
 
 
+@dataclasses.dataclass(frozen=True, repr=False)
```

Comment:
Does the test fail if we do not decorate this with `@dataclass`?

Reply:
Yeah:
```
    _LEAF_SPEC = LeafSpec()
                 ^^^^^^^^^^
TypeError: TreeSpec.__init__() missing 3 required positional arguments: 'type', 'context', and 'children_specs'
```

So the requirements here are:
1. `LeafSpec()` works
2. `dataclasses.fields(LeafSpec())` has `init=False`
3. `LeafSpec` subclasses `TreeSpec`

(3) conflicts with (2), thereby the extra `dataclasses.field(default=None, init=False)` below, which further requires an extra `dataclasses.dataclass` annotation above to work out.

Reply:
Good call; we need the change to `LeafSpec` for #146367, I'll move the change to that PR.